File: Backend/src/main/java/com/shubham/FlightManagementSystem/service/FlightService.java
package com.shubham.FlightManagementSystem.service;

// import com.shubham.FlightManagementSystem.entity.Airport;
import com.shubham.FlightManagementSystem.entity.Flight;
import com.shubham.FlightManagementSystem.entity.Flight.FlightStatus;
import com.shubham.FlightManagementSystem.repository.FlightRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

// import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class FlightService {
    
    @Autowired
    private FlightRepository flightRepository;
    
    @Autowired
    private AirportService airportService;
    
    public List<Flight> getAllFlights() {
        return flightRepository.findAll();
    }
    
    public Optional<Flight> getFlightById(Long id) {
        return flightRepository.findById(id);
    }
    
    public Optional<Flight> getFlightByNumber(String flightNumber) {
        return flightRepository.findByFlightNumber(flightNumber);
    }
    
    public List<Flight> getFlightsByDepartureAirport(String airportCode) {
        return flightRepository.findByDepartureAirportCode(airportCode);
    }
    
    public List<Flight> getFlightsByArrivalAirport(String airportCode) {
        return flightRepository.findByArrivalAirportCode(airportCode);
    }
    
    public List<Flight> searchFlights(String departureCode, String arrivalCode, 
                                    LocalDateTime departureDate, Integer seats) {
        return flightRepository.searchFlights(departureCode, arrivalCode, departureDate, seats);
    }
    
    public List<Flight> getAvailableFlights() {
        return flightRepository.findAvailableFlights();
    }
    
    public List<Flight> getFlightsByStatus(FlightStatus status) {
        return flightRepository.findByStatus(status);
    }
    
    public Flight saveFlight(Flight flight) {
        // Validate airports exist
        if (!airportService.airportExists(flight.getDepartureAirport().getCode())) {
            throw new RuntimeException("Departure airport not found: " + flight.getDepartureAirport().getCode());
        }
        if (!airportService.airportExists(flight.getArrivalAirport().getCode())) {
            throw new RuntimeException("Arrival airport not found: " + flight.getArrivalAirport().getCode());
        }
        
        // Set initial available seats to total seats
        flight.setAvailableSeats(flight.getTotalSeats());
        
        return flightRepository.save(flight);
    }
    
    public Flight updateFlight(Long id, Flight flightDetails) {
        Optional<Flight> optionalFlight = flightRepository.findById(id);
        if (optionalFlight.isPresent()) {
            Flight flight = optionalFlight.get();
            flight.setFlightNumber(flightDetails.getFlightNumber());
            flight.setDepartureAirport(flightDetails.getDepartureAirport());
            flight.setArrivalAirport(flightDetails.getArrivalAirport());
            flight.setDepartureTime(flightDetails.getDepartureTime());
            flight.setArrivalTime(flightDetails.getArrivalTime());
            flight.setAircraftType(flightDetails.getAircraftType());
            flight.setTotalSeats(flightDetails.getTotalSeats());
            flight.setPrice(flightDetails.getPrice());
            flight.setStatus(flightDetails.getStatus());
            return flightRepository.save(flight);
        }
        throw new RuntimeException("Flight not found with id: " + id);
    }
    
    public void deleteFlight(Long id) {
        flightRepository.deleteById(id);
    }
    
    public boolean reserveSeats(Long flightId, Integer numberOfSeats) {
        Optional<Flight> optionalFlight = flightRepository.findById(flightId);
        if (optionalFlight.isPresent()) {
            Flight flight = optionalFlight.get();
            if (flight.getAvailableSeats() >= numberOfSeats) {
                flight.setAvailableSeats(flight.getAvailableSeats() - numberOfSeats);
                flightRepository.save(flight);
                return true;
            }
        }
        return false;
    }
    
    public void releaseSeats(Long flightId, Integer numberOfSeats) {
        Optional<Flight> optionalFlight = flightRepository.findById(flightId);
        if (optionalFlight.isPresent()) {
            Flight flight = optionalFlight.get();
            int newAvailableSeats = flight.getAvailableSeats() + numberOfSeats;
            if (newAvailableSeats <= flight.getTotalSeats()) {
                flight.setAvailableSeats(newAvailableSeats);
                flightRepository.save(flight);
            }
        }
    }
    
    public boolean flightExists(String flightNumber) {
        return flightRepository.existsByFlightNumber(flightNumber);
    }
}

File: Frontend/app/services/flight.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Flight, FlightStatus } from '../models/flight.model';

@Injectable({
  providedIn: 'root'
})
export class FlightService {
  private apiUrl = 'http://localhost:8080/api/flights';

  constructor(private http: HttpClient) { }

  getAllFlights(): Observable<Flight[]> {
    return this.http.get<Flight[]>(this.apiUrl);
  }

  getFlightById(id: number): Observable<Flight> {
    return this.http.get<Flight>(`${this.apiUrl}/${id}`);
  }

  getFlightByNumber(flightNumber: string): Observable<Flight> {
    return this.http.get<Flight>(`${this.apiUrl}/number/${flightNumber}`);
  }

  getFlightsByDepartureAirport(airportCode: string): Observable<Flight[]> {
    return this.http.get<Flight[]>(`${this.apiUrl}/departure/${airportCode}`);
  }

  getFlightsByArrivalAirport(airportCode: string): Observable<Flight[]> {
    return this.http.get<Flight[]>(`${this.apiUrl}/arrival/${airportCode}`);
  }

  getAvailableFlights(): Observable<Flight[]> {
    return this.http.get<Flight[]>(`${this.apiUrl}/available`);
  }

  getFlightsByStatus(status: FlightStatus): Observable<Flight[]> {
    return this.http.get<Flight[]>(`${this.apiUrl}/status/${status}`);
  }

  searchFlights(departureCode: string, arrivalCode: string, departureDate: string, seats: number = 1): Observable<Flight[]> {
    const params = new HttpParams()
      .set('departureCode', departureCode)
      .set('arrivalCode', arrivalCode)
      .set('departureDate', departureDate)
      .set('seats', seats.toString());
    
    return this.http.get<Flight[]>(`${this.apiUrl}/search`, { params });
  }

  createFlight(flight: Flight): Observable<Flight> {
    return this.http.post<Flight>(this.apiUrl, flight);
  }

  updateFlight(id: number, flight: Flight): Observable<Flight> {
    return this.http.put<Flight>(`${this.apiUrl}/${id}`, flight);
  }

  deleteFlight(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  updateFlightStatus(id: number, status: FlightStatus): Observable<Flight> {
    return this.http.put<Flight>(`${this.apiUrl}/${id}/status`, null, {
      params: { status }
    });
  }
}
